### PR TITLE
instancetype: Handle VirtualMachineInstancetypeSpecRevision without APIVersion

### LINKED ISSUE #9261: Failed to apply instancetype: failed to decode object in ControllerRevision: Object 'Kind' is missing in
Hello everyone!

**What happened**:
I've upgraded kubevirt from version v0.57.1 to v0.58.1 and after that i tried to restart virtualmachine, but it's failed to start with error message:
```
  Warning  FailedCreate  11m (x37 over 5h33m)  virtualmachine-controller  Error creating virtual machine instance: Failed to apply instancetype: failed to decode object in ControllerRevision: Object 'Kind' is missing in '{"apiVersion":"","spec":"eyJjcHUiOnsiZ3Vlc3QiOjZ9LCJtZW1vcnkiOnsiZ3Vlc3QiOiIzMkcifX0="}'
```
I managed to solve this issue by removing the old controllerRevision/instancetype resource and revisionName from  the virtualmachine spec manually after that kubevirt created a new ControllerRevision/instancetype with the correct format and the virtualmachine started without issue. I would expect that kubevirt will handle this error properly and migrate resource during  the upgrade.

Instancetype from vm spec
```
spec:
  instancetype:
    kind: VirtualMachineInstancetype
    name: m4.2xlarge
    revisionName: m4.2xlarge-8fffb940-1206-41dd-9529-70cf1faba822-1
```

Old ControllerRevision/instancetype resource format:
```
apiVersion: apps/v1
data:
  apiVersion: ""
  spec: eyJjcHUiOnsiZ3Vlc3QiOjZ9LCJtZW1vcnkiOnsiZ3Vlc3QiOiIzMkcifX0=
kind: ControllerRevision
metadata:
  name: m4.2xlarge-8fffb940-1206-41dd-9529-70cf1faba822-1
```

New ControllerRevision/instancetype resource format:
```
apiVersion: apps/v1
kind: ControllerRevision
metadata:
  name: m4.2xlarge-8fffb940-1206-41dd-9529-70cf1faba822-1
data:
  apiVersion: instancetype.kubevirt.io/v1alpha2
  kind: VirtualMachineInstancetype
  metadata:
    name: m4.2xlarge
  spec:
    cpu:
      guest: 6
    memory:
      guest: 32G
```

**Environment**:
- KubeVirt version (use `virtctl version`): v0.58.1
- Kubernetes version (use `kubectl version`): v1.23.10

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype

**What this PR does / why we need it**:

As set out in issue #9261 there have been reports of older VirtualMachineInstancetypeSpecRevision objects being stored in ControllerRevisions without APIVersion set. This is likely due to a known client-go issue [1] stripping TypeMeta from returned objects.

Given the limited support for v1alpha1 and the fact that we moved away from VirtualMachineInstancetypeSpecRevision to complete objects being captured in ControllerRevisions by v1alpah2 this change simply works around this issue by ignoring the value of APIVersion. This should allow the conversion to complete correctly until we upgrade these ControllerRevisions in the future.

[1] https://github.com/kubernetes/client-go/issues/541

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9261 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/instancetype/instancetype_test.go [hunk 1]
```diff
@@ -423,6 +466,41 @@ var _ = Describe("Instancetype and Preferences", func() {
...
+				_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), controllerRevision, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+					Name:         fakeInstancetype.Name,
+					RevisionName: controllerRevision.Name,
+					Kind:         apiinstancetype.SingularResourceName,
+				}
+
+				foundInstancetypeSpec, err := instancetypeMethods.FindInstancetypeSpec(vm)
```
Comment (0xFelix) [inline pkg/instancetype/instancetype_test.go:500]: Same as above.

File: pkg/instancetype/instancetype_test.go [hunk 2]
```diff
@@ -280,6 +285,41 @@ var _ = Describe("Instancetype and Preferences", func() {
...
+				_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), controllerRevision, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+					Name:         clusterInstancetype.Name,
+					RevisionName: controllerRevision.Name,
+					Kind:         apiinstancetype.ClusterSingularResourceName,
+				}
+
+				foundInstancetypeSpec, err := instancetypeMethods.FindInstancetypeSpec(vm)
```
Comment (0xFelix) [inline pkg/instancetype/instancetype_test.go:319]: Can you somehow assert that the `apiVersion` is empty?

File: pkg/instancetype/instancetype_test.go [hunk 3]
```diff
@@ -305,6 +345,9 @@ var _ = Describe("Instancetype and Preferences", func() {
 						CPU: instancetypev1alpha2.CPUInstancetype{
 							Guest: uint32(2),
 						},
+						Memory: instancetypev1alpha2.MemoryInstancetype{
```
Comment (0xFelix) [inline pkg/instancetype/instancetype_test.go:348]: Same as above.

File: pkg/instancetype/instancetype_test.go [hunk 4]
```diff
@@ -161,6 +163,9 @@ var _ = Describe("Instancetype and Preferences", func() {
...
```
Comment (0xFelix) [inline pkg/instancetype/instancetype_test.go:166]: Is this change needed?

File: pkg/instancetype/instancetype_test.go (hunk 2 above)
Comment (lyarwood) [reply pkg/instancetype/instancetype_test.go:319]: L293 leaves it empty within the `VirtualMachineInstancetypeSpecRevision` stashed within the `ControllerRevision`. I can't think why we'd need to assert that again here?

File: pkg/instancetype/instancetype_test.go (hunk 4 above)
Comment (lyarwood) [reply pkg/instancetype/instancetype_test.go:166]: Yeah the `Equals` on L321 wasn't happy without this required field also being set.

File: pkg/instancetype/instancetype_test.go (hunk 2 above)
Comment (0xFelix) [reply pkg/instancetype/instancetype_test.go:319]: IMO explicitly stating something is cleaner than implicitly relying on it. Yes it is left empty in L293 but it is not obvious at first glance. At least for me.

File: pkg/instancetype/instancetype_test.go (hunk 4 above)
Comment (0xFelix) [reply pkg/instancetype/instancetype_test.go:166]: ACK

File: pkg/instancetype/instancetype_test.go (hunk 3 above)
Comment (0xFelix) [reply pkg/instancetype/instancetype_test.go:348]: ACK

File: pkg/instancetype/instancetype_test.go (hunk 2 above)
Comment (lyarwood) [reply pkg/instancetype/instancetype_test.go:319]: I agree but it's not a property of `foundInstancetypeSpec` that we can assert, to assert that the field isn't set on the stashed `VirtualMachineInstancetypeSpecRevision` within the `ControllerRevision` that we've built above just seems redundant in this context.

Comment (0xFelix) [reply pkg/instancetype/instancetype_test.go:319]: Ok, then maybe add a comment to at least indicate that the apiVersion is left out on purpose?

**Review by 0xFelix — APPROVED** [review]
Thanks!

/lgtm
/retest

**Comment by vladikr** [comment]
/approve
Sorry for the delay. it took me a while to wrap my head around this.

**Comment by vladikr** [comment]
/approve
Sorry for the delay. it took me a while to wrap my head around this.

**Pull request merged**